source/hwIo/ioThread.py

```diff
@@ -10,176 +10,245 @@
-
-		self._apcReferences[apcUuid] = apc
-		return apc
+		self._apcStore[internalParam] = (func or reference, param)
+		return internalParam
 
 	def queueAsApc(
 			self,
-			func: typing.Callable[[int], None],
+			func: ApcT,
 			param: int = 0
 	):
-		"""safely queues an Asynchronous Procedure Call (APC) created from a python function.
-		The generated APC is saved in a cache on the IoThread instance
-		and automatically cleaned when the call is complete.
-		Note that starting from NVDA 2024.1, the wrapped python function will be weakly referenced,
+		"""safely queues a Python function call as an Asynchronous Procedure Call (APC).
+		The function and param are saved in a store on the IoThread instance.
+		When our internal APC executes the function, the entry will be popped from the store.
+		Note that starting from NVDA 2024.1, the queued python function will be weakly referenced,
 		therefore the caller should keep a reference to the python function.
-		@param func: The function to be wrapped in an APC.
+		@param func: The function to be called in an APC.
 		@param param: The parameter passed to the APC when called.
 		"""
-		apc = self._getApc(func, param, _alwaysReferenceWeakly=False)
-		ctypes.windll.kernel32.QueueUserAPC(apc, self.handle, param)
+		internalParam = self._registerToCallAsApc(func, param, _alwaysReferenceWeakly=False)
+		ctypes.windll.kernel32.QueueUserAPC(self._internalApc, self.handle, internalParam)
 
 	def setWaitableTimer(
 			self,
 			handle: typing.Union[int, ctypes.wintypes.HANDLE],
 			dueTime: int,
-			func: typing.Callable[[int], None],
+			func: ApcT,
 			param: int = 0
 	):
-		""""Safe wrapper around winKernel.setWaitableTimer to ensure that the queued APC
-		is available when called.
-		The generated APC is saved in a cache on the IoThread instance
-		and automatically cleaned when the call is complete.
-		The wrapped python function is weakly referenced, therefore the caller should
+		""""Safe wrapper around winKernel.setWaitableTimer that uses an internal APC.
+		A weak reference to the function and its param are saved in a store on the IoThread instance.
+		When our internal APC executes the function, the entry will be popped from the store.
+		Note that as the python function is weakly referenced, the caller should
 		keep a reference to the python function.
 		@param handle: A handle to the timer object.
 		@param dueTime: Relative time (in miliseconds).
 		@param func: The function to be executed when the timer elapses.
 		@param param: The parameter passed to the APC when called.
 		"""
-		apc = self._getApc(func, param)
-		winKernel.setWaitableTimer(handle, dueTime, completionRoutine=apc, arg=param)
+		internalParam = self._registerToCallAsApc(func, param)
+		winKernel.setWaitableTimer(
+			handle,
+			dueTime,
+			completionRoutine=self._internalApc,
+			arg=internalParam
+		)
 
-	def getCompletionRoutine(
+	def queueAsCompletionRoutine(
```

Comment:
changing this name would be an API breaking change, correct?

Reply:
No. This was added in https://github.com/nvaccess/nvda/pull/14627 , so introduced in 2023.2

Reply:
Perhaps this should also be marked as private to avoid API breaks in the future

Reply:
The use case to queue a completion routine from an add-on is perfectly valid, I think. In fact, I'll be likely to use it from [nvdaRD](https://github.com/leonardder/nvdard)